app/controllers/collections_controller.rb

```diff
@@ -189,4 +233,218 @@ def respond_with_error(message, status_code)
+      validate_item(item, collection, corpus_dir)
+    end
+  end
+
+  # Iterates over the uploaded files and JSON document content and returns a list of document identifiers/filenames
+  def get_document_identifiers(uploaded_files, items_metadata)
+    document_identifiers = []
+    items_metadata.each do |item|
+      unless item["documents"].nil?
+        item["documents"].each do |document|
+          document_identifiers.push(document["identifier"])
+        end
+      end
+    end
+    unless uploaded_files.nil?
+      uploaded_files.each do |uploaded_file|
+        document_identifiers.push(uploaded_file.original_filename)
+      end
+    end
+    document_identifiers
+  end
+
+  # Validates that each of the document identifiers are unique
+  def validate_document_identifiers(document_identifiers)
+    duplicate_id = document_identifiers.detect{|identifier| document_identifiers.count(identifier) > 1}
+    raise ResponseError.new(412), "The identifier \"#{duplicate_id}\" is used for multiple documents" unless duplicate_id.nil?
+  end
+
+  # Validates the item doesn't exist in the collection and validates any document metadata with the item
+  def validate_item(item_metadata, collection, corpus_dir)
+    existing_item = Item.find_by_handle("#{collection.name}:#{item_metadata["identifier"]}")
+    if existing_item
+      raise ResponseError.new(412), "The item #{item_metadata["identifier"]} already exists in the collection #{collection.name}"
+    end
+    unless item_metadata["documents"].nil?
+      item_metadata["documents"].each do |document|
+        validate_document(document, item_metadata, collection, corpus_dir)
+      end
+    end
+  end
+
+  # Validates required document parameters present and document file isn't already in the collection directory
+  def validate_document(document_metadata, item_metadata, collection, corpus_dir)
+    if document_metadata["identifier"].nil? or document_metadata["content"].nil?
+      err_message = "identifier missing from document" if document_metadata["identifier"].nil?
+      err_message = "content missing from document #{document_metadata["identifier"]}" if document_metadata["content"].nil?
+      err_message << " for item #{item_metadata["identifier"]}"
+      raise ResponseError.new(400), "#{err_message}"
+    end
+    validate_new_document_file(corpus_dir, document_metadata["identifier"], collection)
+  end
+
+  # Validates that the document file to be created/uploaded doesn't already exist in the collection directory
+  def validate_new_document_file(corpus_dir, file_basename, collection)
+    absolute_filename = File.join(corpus_dir, file_basename)
+    if File.exists? absolute_filename
+      raise ResponseError.new(412), "The file \"#{file_basename}\" has already been uploaded to the collection #{collection.name}"
+    end
+  end
+
+  # Validates each of the uploaded files
+  def validate_files(uploaded_files, collection, corpus_dir)
+    unless uploaded_files.nil?
```

Comment:
could possible avoid this nil check if `cleanse_params` replaces a `params[:file]` param with an empty array